src/HTSPConnection.cpp

```diff
@@ -187,6 +194,32 @@ void CHTSPConnection::OnWake ( void )
   m_suspended = false;
 }
 
+void CHTSPConnection::SetState ( PVR_CONNECTION_STATE state )
+{
+  PVR_CONNECTION_STATE prevState(PVR_CONNECTION_STATE_UNKNOWN);
+  PVR_CONNECTION_STATE newState(PVR_CONNECTION_STATE_UNKNOWN);
+
+  {
+    CLockObject lock(m_mutex);
+
+    /* No notification if no state change or while suspended. */
+    if (m_state != state && !m_suspended)
+    {
+      prevState = m_state;
+      newState  = state;
+      m_state   = newState;
+
+      Logger::Log(LogLevel::LEVEL_DEBUG, "connection state change (%d -> %d)", prevState, newState);
+    }
+  }
+
+  if (prevState != newState)
+  {
+    /* Notify connection state change (callback!) */
+    PVR->ConnectionStateChange(GetServerString().c_str(), newState, NULL);
```

Comment:
No, callback must not called with mutex locked. https://github.com/kodi-pvr/pvr.hts/pull/195#discussion-diff-52122902R203